### ISSUE TITLE
How to display which instances are in NeedsUpdate state?

### ISSUE THREAD

**Issue by hairyhenderson** [kubernetes/kops#7902]
**1. Describe IN DETAIL the feature/behavior/change you would like to see.**

When running `kops rolling-update cluster` I might see output like:

```
NAME                    STATUS          NEEDUPDATE      READY   MIN     MAX     NODES
bastions                Ready           0               1       1       1       0
master-us-east-1a       Ready           0               1       1       1       1
master-us-east-1b       Ready           0               1       1       1       1
master-us-east-1c       Ready           0               1       1       1       1
nodes                   NeedsUpdate     2               1       2       40      2

Must specify --yes to rolling-update.
```

This is great, and if I want to continue and actually roll the cluster, this is enough.

However, I have a use-case where instead of rolling one of the InstanceGroups, I want to first see _which_ nodes specifically need updating.

**2. Feel free to provide a design supporting your feature request.**

A `kops rolling-update cluster -o json` (or `-o yaml`) may be enough - I would envision it returning the full details of _which_ nodes need to be updated, in a format that I can easily post-process with other tooling.

**Comment by johngmyers** [comment]
More detail on the use case would help in evaluating the value and appropriateness of the proposed feature.

/area rolling-update

**Comment by hairyhenderson** [comment]
@johngmyers sure - in my particular use case, I don't actually want the nodes to be actively rolled by kops, instead I want to be able to cordon the nodes needing rolling, then let cluster-autoscaler spin up new nodes as necessary over a longer period of time. The reason is due to a stateful workload that doesn't (yet) support shifting user sessions transparently between replicas.

Beyond that, it just seems like the `kops rolling-update cluster` command should have a more machine-readable output as an option, rather than only outputting data in a hard-to-parse tabular form. This would bring it more in line with commands like `kops get cluster -o yaml` or `kops get instancegroups ... -o yaml` as well.

**Comment by johngmyers** [comment]
@hairyhenderson Thanks. I can see how two features I'm hoping to add, cordonAllNeedUpdate and maxUnavailable, could combine to solve your use case.

Could you expand a bit on why `terminationGracePeriodSeconds` and `PodDisruptionBudget` aren't sufficient for your workload? How do you get the workload shifted off the cordoned nodes?

**Comment by hairyhenderson** [comment]
@johngmyers `terminationGracePeriodSeconds` is actually _partly_ sufficient for my workload, in some cases, but in some cases it can be many hours or sometimes days before some pods are ready to be finally terminated.

We use a CI/CD system to do our deploys, and it's not desirable to have a deploy job that takes many days to roll the cluster, especially when the rolling update happens node-by-node.

We also use the cluster-autoscaler, and when we have all of the nodes that are in a `NeedsUpdate` state cordoned off, we can instead take an approach of "letting nature take its course" - i.e. as deploys and HPA scale-outs happen for non-stateful services, new nodes are provisioned by cluster-autoscaler, and over time the old nodes empty themselves out.

Now obviously this approach could take forever, and that's no good either - especially when the updates we want to push include time-sensitive security updates. In those cases, it's useful to be able to do the deploy at any time during the day, and then as an after-hours activity (when most users are inactive), we can `kubectl drain` the cordoned nodes, and bring them down manually. Or optionally just drain them and let cluster-autoscaler reap the nodes with very little workload left, as usual.

In my mind, it's a bit of a "gentler" approach to rolling clusters that have highly stateful workloads which can take hours to roll.

I do have a partial workaround for this now -  basically just `sed`ing the output from the command, then cordoning each instance group (with `kubectl cordon -l "kops.k8s.io/instancegroup==$group"`), but it's not ideal since it's possible that some nodes in the instance group _don't_ need updating.

**Comment by johngmyers** [comment]
@hairyhenderson I think #8021 might go a long way towards supporting your use case. Put your special workloads under one or more `PodDisruptionBudget`s that can't be satisfied, configure the instancegroup to cordon all nodes needing update, then let the rolling update hang or time out when it is then unable to evict one of those pods.

The issue I see would be that PR's protection against cordoning all nodes when the new instance spec produces nodes that fail validation. You wouldn't want all your nodes cordoned when cluster-autoscaler can't create usable new ones. You'd probably need some way to surge by one node in order to create a new node that proves the spec is good.

**Comment by hairyhenderson** [comment]
@johngmyers Thanks! I've learned some things (the hard way) since posting this - turns out that in AWS, nodes that are cordoned will be marked as unhealthy to the ELB and taken out of its group of instances. This is a problem when _all_ the nodes are cordoned, since all of a sudden you have a 100% site outage 😅

I've moved to a different approach, which is to taint the nodes with a specific `NoSchedule` taint, which unlike cordoning leaves the nodes in the ELB.

However, I am seeing issues where `cluster-autoscaler` is a bit sluggish and doesn't immediately scale out.

> You'd probably need some way to surge by one node in order to create a new node that proves the spec is good.

absolutely, and in my case I'm finding I need to scale out instance groups by hand anyway. It's a royal pain TBH 😉

Bottom line - I still think it would be helpful for `kops` to provide some more consumable information (e.g. a list of nodes in `NeedsUpdate` state) so that we can experiment and figure out what the right way around this is.

**Comment by johngmyers** [comment]
I am reluctant to expose internal implementation details through a public API like that. For example, I suspect that in order to implement surging I will need to add a third node state.

I suspect your use case can be addressed by a combination of PodDisruptionBudgets, preventing scheduling, and surging. The latter two I intend to implement.

**Comment by hairyhenderson** [comment]
> I am reluctant to expose internal implementation details through a public API like that.

How is the list of nodes that needs updating an "internal implementation detail?" It seems to me like just a greater verbosity level for the command...

**Comment by johngmyers** [comment]
A likely implementation of surging is to detach instances from the autoscaling group. This can lead to nodes that need to be brought into conformance despite matching the current specification and which could conceivably be brought into conformance with less disruption than draining and deleting.

As far as my own priorities go, I'm first going to be working on improved rolling update strategies, such as surging. I suspect these might address your use case. Others might have different priorities.

**Comment by johngmyers** [comment]
@hairyhenderson Now that surging has landed, I believe it and `PodDisruptionBudget` could solve your use case. Give your workloads appropriate `PodDisruptionBudget`s and set `maxSurge` to `"100%"` (or less if you want to go more slowly).

**Comment by hairyhenderson** [comment]
@johngmyers thanks - which version of kops is your surging feature available in? I don't see #8021 mentioned in any of the release notes.

**Comment by johngmyers** [comment]
It's in the unreleased master branch.

**Comment by hairyhenderson** [comment]
Ah, thanks @johngmyers - I'll look forward that being released!

**Comment by hairyhenderson** [comment]
Still relevant...

/remove-lifecycle rotten
